File: src/lib/api/knowledge.ts
import { invoke } from "@tauri-apps/api/core";
import type { KnowledgeEntry, KnowledgeFolder } from "./types";

// ─── Folders ────────────────────────────────────────────────────────────────

export async function getKnowledgeFolders(): Promise<KnowledgeFolder[]> {
  return invoke<KnowledgeFolder[]>("get_knowledge_folders");
}

export async function createKnowledgeFolder(
  name: string,
  parentId?: string | null,
): Promise<KnowledgeFolder> {
  return invoke<KnowledgeFolder>("create_knowledge_folder", {
    name,
    parentId: parentId ?? null,
  });
}

export async function renameKnowledgeFolder(
  id: string,
  name: string,
): Promise<KnowledgeFolder> {
  return invoke<KnowledgeFolder>("rename_knowledge_folder", { id, name });
}

export async function deleteKnowledgeFolder(id: string): Promise<void> {
  return invoke<void>("delete_knowledge_folder", { id });
}

// ─── Entries ────────────────────────────────────────────────────────────────

export async function getKnowledgeEntries(
  folderId?: string | null,
): Promise<KnowledgeEntry[]> {
  return invoke<KnowledgeEntry[]>("get_knowledge_entries", {
    folderId: folderId ?? null,
  });
}

export async function getAllKnowledgeEntries(): Promise<KnowledgeEntry[]> {
  return invoke<KnowledgeEntry[]>("get_all_knowledge_entries");
}

export async function getKnowledgeEntry(id: string): Promise<KnowledgeEntry> {
  return invoke<KnowledgeEntry>("get_knowledge_entry", { id });
}

export async function createKnowledgeEntry(params: {
  title: string;
  content: string;
  description?: string;
  folderId?: string | null;
  featureId?: string | null;
}): Promise<KnowledgeEntry> {
  return invoke<KnowledgeEntry>("create_knowledge_entry", {
    title: params.title,
    content: params.content,
    description: params.description ?? null,
    folderId: params.folderId ?? null,
    featureId: params.featureId ?? null,
  });
}

export async function updateKnowledgeEntry(params: {
  id: string;
  title?: string;
  content?: string;
  description?: string;
  folderId?: string | null;
  featureId?: string | null;
}): Promise<KnowledgeEntry> {
  return invoke<KnowledgeEntry>("update_knowledge_entry", {
    id: params.id,
    title: params.title ?? null,
    content: params.content ?? null,
    description: params.description ?? null,
    folderId: params.folderId !== undefined ? params.folderId : null,
    featureId: params.featureId !== undefined ? params.featureId : null,
  });
}

export async function deleteKnowledgeEntry(id: string): Promise<void> {
  return invoke<void>("delete_knowledge_entry", { id });
}

File: src-tauri/src/commands/knowledge.rs
use tauri::State;
use crate::db;
use crate::AppState;

// ─── Folder commands ────────────────────────────────────────────────────────

#[tauri::command]
pub fn get_knowledge_folders(
    state: State<'_, AppState>,
) -> Result<Vec<db::knowledge::KnowledgeFolder>, String> {
    let conn = state.db.lock().map_err(|e| e.to_string())?;
    db::knowledge::list_folders(&conn)
}

#[tauri::command]
pub fn create_knowledge_folder(
    state: State<'_, AppState>,
    name: String,
    parent_id: Option<String>,
) -> Result<db::knowledge::KnowledgeFolder, String> {
    let conn = state.db.lock().map_err(|e| e.to_string())?;
    db::knowledge::create_folder(&conn, &name, parent_id.as_deref())
}

#[tauri::command]
pub fn rename_knowledge_folder(
    state: State<'_, AppState>,
    id: String,
    name: String,
) -> Result<db::knowledge::KnowledgeFolder, String> {
    let conn = state.db.lock().map_err(|e| e.to_string())?;
    db::knowledge::rename_folder(&conn, &id, &name)
}

#[tauri::command]
pub fn delete_knowledge_folder(
    state: State<'_, AppState>,
    id: String,
) -> Result<(), String> {
    let conn = state.db.lock().map_err(|e| e.to_string())?;
    db::knowledge::delete_folder(&conn, &id)
}

// ─── Entry commands ─────────────────────────────────────────────────────────

#[tauri::command]
pub fn get_knowledge_entries(
    state: State<'_, AppState>,
    folder_id: Option<String>,
) -> Result<Vec<db::knowledge::KnowledgeEntry>, String> {
    let conn = state.db.lock().map_err(|e| e.to_string())?;
    db::knowledge::list_entries_in_folder(&conn, folder_id.as_deref())
}

#[tauri::command]
pub fn get_all_knowledge_entries(
    state: State<'_, AppState>,
) -> Result<Vec<db::knowledge::KnowledgeEntry>, String> {
    let conn = state.db.lock().map_err(|e| e.to_string())?;
    db::knowledge::list_entries(&conn)
}

#[tauri::command]
pub fn get_knowledge_entry(
    state: State<'_, AppState>,
    id: String,
) -> Result<db::knowledge::KnowledgeEntry, String> {
    let conn = state.db.lock().map_err(|e| e.to_string())?;
    db::knowledge::get_entry(&conn, &id)
}

#[tauri::command]
pub fn create_knowledge_entry(
    state: State<'_, AppState>,
    title: String,
    content: String,
    description: Option<String>,
    folder_id: Option<String>,
    feature_id: Option<String>,
) -> Result<db::knowledge::KnowledgeEntry, String> {
    let conn = state.db.lock().map_err(|e| e.to_string())?;
    db::knowledge::create_entry(
        &conn,
        &title,
        &content,
        description.as_deref(),
        folder_id.as_deref(),
        feature_id.as_deref(),
    )
}

#[tauri::command]
pub fn update_knowledge_entry(
    state: State<'_, AppState>,
    id: String,
    title: Option<String>,
    content: Option<String>,
    description: Option<String>,
    folder_id: Option<Option<String>>,
    feature_id: Option<Option<String>>,
) -> Result<db::knowledge::KnowledgeEntry, String> {
    let conn = state.db.lock().map_err(|e| e.to_string())?;
    db::knowledge::update_entry(
        &conn,
        &id,
        title.as_deref(),
        content.as_deref(),
        description.as_deref(),
        folder_id.as_ref().map(|o| o.as_deref()),
        feature_id.as_ref().map(|o| o.as_deref()),
    )
}

#[tauri::command]
pub fn delete_knowledge_entry(
    state: State<'_, AppState>,
    id: String,
) -> Result<(), String> {
    let conn = state.db.lock().map_err(|e| e.to_string())?;
    db::knowledge::delete_entry(&conn, &id)
}
